### DISCUSSION TITLE
Updating pivot records

### DISCUSSION THREAD

**Discussion by JBSommeling** [filamentphp/filament#2121] [Help] [answered]
Hi all,

I have a User model with a BelongsToMany relationship to a model called Tale. The relation is shown in filament admin through a TalesRelationManager. So far so good.

I want to access the pivot table's data (TaleUser) through the TalesRelationManager. When using dot notation, displaying the records is not a problem. This also works fine. However, when trying to edit/update these records, this does not work as I get an error:

```
update `tales` set `taleUser` = {"max_progress":"3"}, `tales`.`updated_at` = 2022-04-11 08:41:18 where `id` = 2
```

My code from TalesRelationManager:

```
    public static function form(Form $form): Form
    {
        return $form
            ->schema([
                // @TODO:: try to find a way to make this work. Max_progress will not be updated.
                Forms\Components\TextInput::make('taleUser.max_progress')->required(),
            ]);
    }

    public static function table(Table $table): Table
    {
        return $table
            ->columns([
                Tables\Columns\TextColumn::make('title')->sortable(),
                Tables\Columns\TextColumn::make('taleUser.max_progress')->label('Max progress')->sortable(),
            ])
            ->filters([
                //
            ]);
    }

    public static function attachForm(Form $form): Form
    {
        return $form
            ->schema([
                static::getAttachFormRecordSelect(),
                // @TODO:: try to find a way to make this work. Max_progress will not be updated.
                Forms\Components\TextInput::make('taleUser.max_progress')->required(),
            ]);
    }
```

I was wondering what I am doing wrong. Could anyone please help me? :)
Or is this not yet possible in filament 2.x?

Edit:
Thanks to @ralphjsmit I figured out what is wrong.

When debugging: 

```
Forms\Components\TextInput::make('taleUser.max_progress')
                    ->saveRelationshipsUsing(function (Model $record, string $state) {
                        dd($record, $state);
                        $record->taleUser->max_progress = $state;
                    })
                    ->dehydrated(false)
                    ->required(),
```

It says the relationships are null. Which is weird because the Tale model has:
```
class Tale extends Model
{
    use HasFactory,

    protected $guarded = [];

    // Relations
    public function users(): BelongsToMany
    {
        return $this->belongsToMany(User::class)->using(TaleUser::class);
    }
    
    public function taleUser(): HasOne
    {
        return $this->hasOne(TaleUser::class, 'tale_id', 'id');
    }
}
```
There is something wrong with the relationship? It is not seeing the given pivot taleUser? In the project, other than filament, the pivot relation is seen and loaded without a problem.

**Comment by SebastiaanKloos** [comment]
@JBSommeling Filament knows about the pivot columns if you've set it up correctly. Try to change `taleUser.max_progress` to just `max_progress`

**Comment by SebastiaanKloos** [comment]
@JBSommeling Please also take a look at https://filamentphp.com/docs/2.x/admin/resources#pivot-attributes

**Reply by JBSommeling** [reply]
I have read this, thank you. It somehow still does not work?
I have changed it to just `max_progress` Now the error I'm getting is:
```
update
  `tales`
set
  `max_progress` = 4,
  `tales`.`updated_at` = 2022 -04 -11 12: 27: 59
where
  `id` = 2`

```

Filament is still looking for the tales table, instead of tale_user? Any suggestions I could lookout for? :)

**Reply by ryangjchandler** [reply]
You can only edit pivot columns inside of a relation manager, not from the normal Resource form.

**Reply by JBSommeling** [reply]
That is what I'm doing yes. Maybe I was not clear enough. I am editing code from inside the TalesRelationManager. I will edit my original question :)

**Comment by ralphjsmit** [comment, marked as answer]
How about something like this?

```php
Forms\Components\TextInput::make('taleUser.max_progress')
    ->saveRelationshipsUsing(function (Model $record, array $state) {
        // $record->taleUser->max_progress = $state; ...
    })
    ->dehydrated(false)
    ->required(),
```

**Reply by JBSommeling** [reply]
Thank you for your reply. This almost works! I think :) 
When debugging your code: 

```
Forms\Components\TextInput::make('taleUser.max_progress')
                    ->saveRelationshipsUsing(function (Model $record, string $state) {
                        dd($record, $state);
                        $record->taleUser->max_progress = $state;
                    })
                    ->dehydrated(false)
                    ->required(),
```

It says the relationships are null. Which is weird because the Tale model has:
```
class Tale extends Model
{
    use HasFactory,

    protected $guarded = [];

    // Relations
    public function users(): BelongsToMany
    {
        return $this->belongsToMany(User::class)->using(TaleUser::class);
    }
    
    public function taleUser(): HasOne
    {
        return $this->hasOne(TaleUser::class, 'tale_id', 'id');
    }
}
```

Any suggestions?  I will edit my post with this. :)

There is something wrong with the relationship? It is not seeing the given pivot taleUser? In the project, other than filament, the pivot relation is seen and loaded without a problem.

**Reply by ralphjsmit** [reply]
Nice :) 

I'm not 100% sure about the pivot attribute. However, you say that TaleUser is a model to represent the pivot record. But in your example you also have a relationship that calls the taleUser() directly. Shouldn't you only keep one of those relationships?

E.g.

```php
Forms\Components\TextInput::make('taleUser.max_progress')
    ->saveRelationshipsUsing(function (Model $record, array $state) {
        // Determine the user, then update the pivot.
        $record->users()->first()->pivot->update([
            'max_progress' => $state,
        ]);
    })
    ->dehydrated(false)
    ->required(),
```

**Reply by JBSommeling** [reply]
Thanks a lot! 

Not sure really if i am to use only one of those relationships. In my experience it was always possible to define a 'pivot relationship'. Correct me if I am wrong. ;-) 

Your code did help a lot though. I removed the `taleUser.max_progress` in `make()` and just used `max_progress` as it was always trying to update the value from the tales model instead of the pivot. Also, I used the `$record->taleUser()->update()` to force an update. It works! Thank you :) 

This community can expect a donation somewhere in the upcoming days :)

```

public static function form(Form $form): Form
    {
        return $form
            ->schema([
                Forms\Components\TextInput::make('max_progress')
                    ->saveRelationshipsUsing(function (Model $record, string $state) {
                        $record->taleUser()->update([
                            'max_progress' => $state,
                        ]);
                    })
                    ->dehydrated(false)
                    ->required(),
            ]);
    }

**Reply by ralphjsmit** [reply]
Happy to hear that you got it working!
